### PR TITLE
S3966 dispose more

### LINKED ISSUE #209: Rule S3966: Objects should not be disposed more than once
Implements [RSPEC-3966: Objects should not be disposed more than once](https://jira.sonarsource.com/browse/RSPEC-3966)

Contributes to [MMF-963](https://jira.sonarsource.com/browse/MMF-963).

### PR DESCRIPTION
**Pull request by valhristov**
Fix #209

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ObjectsShouldNotBeDisposedMoreThanOnce.cs [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+            public ObjectDisposedPointerCheck(ExplodedGraph explodedGraph)
+                : base(explodedGraph)
+            {
+            }
+
+            public override ProgramState PreProcessInstruction(ProgramPoint programPoint, ProgramState programState)
+            {
+                var instruction = programPoint.Block.Instructions[programPoint.Offset];
+                switch (instruction.Kind())
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ObjectsShouldNotBeDisposedMoreThanOnce.cs]: MAJOR Replace this 'switch' statement with 'if' statements to increase readability. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1301)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ObjectsShouldNotBeDisposedMoreThanOnce.cs [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+        {
+            context.RegisterExplodedGraphBasedAnalysis((e, c) => CheckForNullDereference(e, c));
+        }
+
+        private static void CheckForNullDereference(ExplodedGraph explodedGraph, SyntaxNodeAnalysisContext context)
+        {
+            var objectDisposedCheck = new ObjectDisposedPointerCheck(explodedGraph);
+            explodedGraph.AddExplodedGraphCheck(objectDisposedCheck);
+
+            var nullIdentifiers = new HashSet<IdentifierNameSyntax>();
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ObjectsShouldNotBeDisposedMoreThanOnce.cs]: MINOR Remove this unused 'nullIdentifiers' local variable. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1481)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/ControlFlowGraph/Blocks/UsingFinalizerBlock.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        {
+            UsingStatement = usingStatement;
+
+            if (usingStatement.Declaration != null)
+            {
+                Disposables = usingStatement.Declaration
+                    .Variables
+                    .ToImmutableArray();
+            }
+            else
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/ControlFlowGraph/Blocks/UsingFinalizerBlock.cs]: let's return and get rid of this `else` and its nesting.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/ControlFlowGraph/Blocks/UsingFinalizerBlock.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+                if (identifier != null)
+                {
+                    Disposables = ImmutableArray.Create(identifier);
+                }
+                else
+                {
+                    Disposables = usingStatement.Expression
+                        .DescendantNodesAndSelf()
+                        .OfType<AssignmentExpressionSyntax>()
+                        .Select(a => a.Left as IdentifierNameSyntax)
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/ControlFlowGraph/Blocks/UsingFinalizerBlock.cs]: Maybe better to do `.Select(a => a..Left).OfType<IdentifierNameSyntax>()`. WDYT?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/LiveVariableAnalysis/LiveVariableAnalysis.cs [hunk 5]
```diff
@@ -95,6 +95,17 @@ public static Common.LiveVariableAnalysis Analyze(IControlFlowGraph controlFlowG
                 var foreachNode = (ForEachStatementSyntax)foreachBlock.BranchingNode;
                 ProcessVariableInForeach(foreachNode, assignedInBlock, usedBeforeAssigned);
             }
+
+            var usingFinalizerBlock = block as UsingFinalizerBlock;
+            if (usingFinalizerBlock != null)
+            {
+                var disposableSymbols = usingFinalizerBlock.Disposables
+                    .Select(disposable => semanticModel.GetDeclaredSymbol(disposable) ?? semanticModel.GetSymbolInfo(disposable).Symbol);
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/LiveVariableAnalysis/LiveVariableAnalysis.cs]: No null check?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/LiveVariableAnalysis/LiveVariableAnalysis.cs [hunk 6]
```diff
@@ -95,6 +95,17 @@ public static Common.LiveVariableAnalysis Analyze(IControlFlowGraph controlFlowG
...
             }
+
+            var usingFinalizerBlock = block as UsingFinalizerBlock;
+            if (usingFinalizerBlock != null)
+            {
+                var disposableSymbols = usingFinalizerBlock.Disposables
+                    .Select(disposable => semanticModel.GetDeclaredSymbol(disposable) ?? semanticModel.GetSymbolInfo(disposable).Symbol);
+                foreach (var disposableSymbol in disposableSymbols)
+                {
+                    usedBeforeAssigned.Add(disposableSymbol);
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/LiveVariableAnalysis/LiveVariableAnalysis.cs]: Not sure what's the type of `usedBeforeAssigned` but is there no `.AddRange()`?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ObjectsShouldNotBeDisposedMoreThanOnce.cs [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+        };
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected sealed override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterExplodedGraphBasedAnalysis((e, c) => CheckForNullDereference(e, c));
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ObjectsShouldNotBeDisposedMoreThanOnce.cs]: You can probably write `context.RegisterExplodedGraphBasedAnalysis(CheckForNullDereference)`.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ObjectsShouldNotBeDisposedMoreThanOnce.cs [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+            {
+                var newProgramState = programState;
+
+                var arguments = usingExpression.DescendantNodes()
+                    .OfType<ObjectCreationExpressionSyntax>()
+                    .Where(this.IsStreamDisposingType)
+                    .Select(FirstArgumentOrDefault)
+                    .WhereNotNull();
+
+                foreach (var argument in arguments)
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ObjectsShouldNotBeDisposedMoreThanOnce.cs]: What about another `Select` to have syntax + symbol and then `ToList().ForEach()`?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ObjectsShouldNotBeDisposedMoreThanOnce.cs [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+                var disposables = usingFinalizer.Disposables
+                    .Select(disposable =>
+                    new
+                    {
+                        Disposable = disposable,
+                        Symbol = semanticModel.GetDeclaredSymbol(disposable)
+                            ?? semanticModel.GetSymbolInfo(disposable).Symbol
+                    });
+
+                foreach (var nodeWithSymbol in disposables)
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ObjectsShouldNotBeDisposedMoreThanOnce.cs]: What about `ToList().ForEach()`?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ObjectsShouldNotBeDisposedMoreThanOnce.cs [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+                return newProgramState;
+            }
+
+            private static ArgumentSyntax FirstArgumentOrDefault(ObjectCreationExpressionSyntax objectCreation) =>
+                objectCreation.ArgumentList?.Arguments.FirstOrDefault();
+
+            private bool IsStreamDisposingType(ObjectCreationExpressionSyntax objectCreation) =>
+                semanticModel.GetSymbolInfo(objectCreation.Type)
+                    .Symbol
+                    .GetSymbolType()
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ObjectsShouldNotBeDisposedMoreThanOnce.cs]: Is it ok to call `.GetSymbolType()` on null? Are we sure `.Symbol` is never null?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ObjectsShouldNotBeDisposedMoreThanOnce.cs [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+
+                if (disposableSymbol.HasConstraint(DisposableConstraint.Disposed, programState))
+                {
+                    ObjectDisposed?.Invoke(this, new ObjectDisposedEventArgs(disposableSymbol.Name, instruction.GetLocation()));
+                    return programState;
+                }
+
+                // We should not replace Null constraint because having Disposed constraint
+                // implies having NotNull constraint, which is incorrect.
+                if (!disposableSymbol.HasConstraint(ObjectConstraint.Null, programState))
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ObjectsShouldNotBeDisposedMoreThanOnce.cs]: I would have negated this condition (especially with the comment which made me expect the opposite condition).

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/SonarAnalyzer.CSharp.csproj [hunk 12]
```diff
@@ -91,6 +91,7 @@
     <Compile Include="Helpers\ControlFlowGraph\Blocks\ForeachCollectionProducerBlock.cs" />
     <Compile Include="Helpers\ControlFlowGraph\Blocks\ForInitializerBlock.cs" />
     <Compile Include="Helpers\ControlFlowGraph\Blocks\LockBlock.cs" />
+    <Compile Include="Helpers\ControlFlowGraph\Blocks\UsingFinalizerBlock.cs" />
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/SonarAnalyzer.CSharp.csproj]: Shall we use wildcard at some point?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SymbolicExecution/SymbolicValues/SymbolicValue.cs [hunk 13]
```diff
@@ -228,6 +228,12 @@ public virtual IEnumerable<ProgramState> TrySetConstraint(SymbolicValueConstrain
                 return new[] { currentProgramState };
             }
 
+            var disposableConstraint = constraint as DisposableConstraint;
+            if (disposableConstraint != null)
+            {
+                return new[] { currentProgramState };
+            }
+
             throw new NotSupportedException($"Neither {nameof(BoolConstraint)}, nor {nameof(ObjectConstraint)}, " +
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SymbolicExecution/SymbolicValues/SymbolicValue.cs]: Let's add `DisposableConstraint` here.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/SonarAnalyzer.CSharp.csproj (hunk 12 above)
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/SonarAnalyzer.CSharp.csproj]: Next week, when we switch to VS2017 🗡

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ObjectsShouldNotBeDisposedMoreThanOnce.cs (hunk 10 above)
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ObjectsShouldNotBeDisposedMoreThanOnce.cs]: GetSymbolType is an extension method and returns null when needed

**Pull request merged**